File: ios/Feedback.swift
// feedback-cli Apple client (iOS + macOS) — a tiny, dependency-free drop-in so
// a Swift app can report structured feedback back to its project as a caco
// bead, using the SAME webhook + payload contract as the Rust `feedback-cli`
// crate. Single file: drop into an app/SwiftPM target.
//
//   try await reportFeedback(
//     WebhookConfig(url: "https://host/hooks/global/feedback", token: "..."),
//     FeedbackEvent(kind: .error, component: "sync", summary: "upload failed",
//                   detail: err.localizedDescription, labels: ["ios"]))

import Foundation

public enum FeedbackKind: String { case error, exception, perf, info }
public enum Severity: String { case info, warning, error, critical }

public struct Metric {
    public var name: String; public var value: Double
    public var unit: String?; public var threshold: Double?; public var baseline: Double?
    public init(name: String, value: Double, unit: String? = nil, threshold: Double? = nil, baseline: Double? = nil) {
        self.name = name; self.value = value; self.unit = unit; self.threshold = threshold; self.baseline = baseline
    }
}

public struct FeedbackEvent {
    public var kind: FeedbackKind
    public var component: String
    public var summary: String
    public var detail: String?
    public var severity: Severity?
    public var labels: [String]
    public var fields: [String: String]
    public var fingerprint: String?
    public var project: String?
    public var metric: Metric?
    public var timestampUnixMs: UInt64
    public init(kind: FeedbackKind, component: String, summary: String, detail: String? = nil,
                severity: Severity? = nil, labels: [String] = [], fields: [String: String] = [:],
                fingerprint: String? = nil, project: String? = nil, metric: Metric? = nil,
                timestampUnixMs: UInt64 = UInt64(Date().timeIntervalSince1970 * 1000)) {
        self.kind = kind; self.component = component; self.summary = summary; self.detail = detail
        self.severity = severity; self.labels = labels; self.fields = fields; self.fingerprint = fingerprint
        self.project = project; self.metric = metric; self.timestampUnixMs = timestampUnixMs
    }
}

public struct WebhookConfig {
    public var url: String; public var token: String?; public var payload: String
    public init(url: String, token: String? = nil, payload: String = "caco_bead") {
        self.url = url; self.token = token; self.payload = payload
    }
}

// caco_bead payload (mirrors FeedbackEvent::to_caco_bead in src/lib.rs).
public func toCacoBead(_ e: FeedbackEvent) -> [String: Any] {
    let type = (e.kind == .error || e.kind == .exception) ? "bug" : "task"
    let priority: Int = e.severity == .critical ? 1 : e.severity == .warning ? 3 : e.severity == .info ? 4 : 2
    var labels = ["feedback", "kind:\(e.kind.rawValue)"]
    for l in e.labels where !labels.contains(l) { labels.append(l) }
    var ctx = ["component: \(e.component)"]
    if let s = e.severity { ctx.append("severity: \(s.rawValue)") }
    if let f = e.fingerprint { ctx.append("fingerprint: \(f)") }
    if let p = e.project { ctx.append("project: \(p)") }
    for (k, v) in e.fields.sorted(by: { $0.key < $1.key }) { ctx.append("\(k): \(v)") }
    if let m = e.metric { ctx.append("metric \(m.name): \(m.value)\(m.unit ?? "")") }
    ctx.append("timestamp_unix_ms: \(e.timestampUnixMs)")
    let footer = ctx.joined(separator: "\n")
    let description = (e.detail?.isEmpty == false) ? "\(e.detail!)\n\n\(footer)" : footer
    return ["title": e.summary, "description": description, "type": type, "priority": priority, "labels": labels]
}

// POST the feedback event to the configured caco webhook. Returns true on 2xx.
@discardableResult
public func reportFeedback(_ cfg: WebhookConfig, _ e: FeedbackEvent) async throws -> Bool {
    var req = URLRequest(url: URL(string: cfg.url)!)
    req.httpMethod = "POST"
    req.setValue("application/json", forHTTPHeaderField: "Content-Type")
    if let t = cfg.token { req.setValue("Bearer \(t)", forHTTPHeaderField: "Authorization") }
    req.httpBody = try JSONSerialization.data(withJSONObject: toCacoBead(e))
    let (_, resp) = try await URLSession.shared.data(for: req)
    return (200..<300).contains((resp as? HTTPURLResponse)?.statusCode ?? 0)
}

File: android/Feedback.kt
// feedback-cli Android client (Kotlin) — a tiny, dependency-free drop-in so an
// Android app can report structured feedback back to its project as a caco
// bead, using the SAME webhook + payload contract as the Rust `feedback-cli`
// crate. Single file (java.net.HttpURLConnection only). Call off the main
// thread (e.g. a coroutine on Dispatchers.IO).
//
//   reportFeedback(WebhookConfig("https://host/hooks/global/feedback", token),
//     FeedbackEvent(Kind.ERROR, "sync", "upload failed", detail = e.message, labels = listOf("android")))

package dev.cacophony.feedback

import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL

enum class Kind(val v: String) { ERROR("error"), EXCEPTION("exception"), PERF("perf"), INFO("info") }
enum class Severity(val v: String) { INFO("info"), WARNING("warning"), ERROR("error"), CRITICAL("critical") }

data class Metric(val name: String, val value: Double, val unit: String? = null,
                  val threshold: Double? = null, val baseline: Double? = null)

data class FeedbackEvent(
    val kind: Kind,
    val component: String,
    val summary: String,
    val detail: String? = null,
    val severity: Severity? = null,
    val labels: List<String> = emptyList(),
    val fields: Map<String, String> = emptyMap(),
    val fingerprint: String? = null,
    val project: String? = null,
    val metric: Metric? = null,
    val timestampUnixMs: Long = System.currentTimeMillis(),
)

data class WebhookConfig(val url: String, val token: String? = null, val payload: String = "caco_bead")

// caco_bead payload (mirrors FeedbackEvent::to_caco_bead in src/lib.rs).
fun toCacoBead(e: FeedbackEvent): JSONObject {
    val type = if (e.kind == Kind.ERROR || e.kind == Kind.EXCEPTION) "bug" else "task"
    val priority = when (e.severity) {
        Severity.CRITICAL -> 1; Severity.WARNING -> 3; Severity.INFO -> 4; else -> 2
    }
    val labels = mutableListOf("feedback", "kind:${e.kind.v}")
    e.labels.forEach { if (!labels.contains(it)) labels.add(it) }
    val ctx = mutableListOf("component: ${e.component}")
    e.severity?.let { ctx.add("severity: ${it.v}") }
    e.fingerprint?.let { ctx.add("fingerprint: $it") }
    e.project?.let { ctx.add("project: $it") }
    e.fields.toSortedMap().forEach { (k, v) -> ctx.add("$k: $v") }
    e.metric?.let { ctx.add("metric ${it.name}: ${it.value}${it.unit ?: ""}") }
    ctx.add("timestamp_unix_ms: ${e.timestampUnixMs}")
    val footer = ctx.joinToString("\n")
    val description = if (!e.detail.isNullOrEmpty()) "${e.detail}\n\n$footer" else footer
    return JSONObject(mapOf(
        "title" to e.summary, "description" to description,
        "type" to type, "priority" to priority, "labels" to labels,
    ))
}

/** POST the feedback event to the configured caco webhook. Returns true on 2xx. Call off-main-thread. */
fun reportFeedback(cfg: WebhookConfig, e: FeedbackEvent): Boolean {
    val conn = URL(cfg.url).openConnection() as HttpURLConnection
    return try {
        conn.requestMethod = "POST"
        conn.doOutput = true
        conn.setRequestProperty("Content-Type", "application/json")
        cfg.token?.let { conn.setRequestProperty("Authorization", "Bearer $it") }
        conn.outputStream.use { it.write(toCacoBead(e).toString().toByteArray()) }
        conn.responseCode in 200..299
    } finally {
        conn.disconnect()
    }
}
